### PR TITLE
Prometheus: Fix query editor crash in response to invalid query

### LINKED ISSUE #94219: Prometheus: Query editor crashes in response to invalid query
## What's happening?

When querying Prometheus in Explore with the following work-in-progress query:
```
histogram_count(0,sum by (le) (increase(duration_seconds{deployment_environment=~\"someEnv\", span_kind=~\"SPAN_KIND_SERVER|SPAN_KIND_CONSUMER\", job=\"someService\"} [$__range])))
```
, the query editor crashes:

Image

## What's the expected behavior?

Instead of crashing, the query editor should provide users with feedback about issues with the query:

Image

## Why is this happening?

In `packages/grafana-prometheus/src/querybuilder/operationUtils.ts`:

https://github.com/grafana/grafana/blob/4623a6471bf77036ea3f0f17b0ff4e94272bb01a/packages/grafana-prometheus/src/querybuilder/operationUtils.ts#L96-L105

, it's possible for `def.params` to be an empty array. In that case, the check for `paramDef.type === 'string'` is problematic.

## How to fix?

Add some optional chaining to the `paramDef.type` check. A higher-level unit or integration test could help to guard against future regressions.

While we're at it, we could marginally improve clarity & performance by using string interpolation instead of concatenation when returning the quote-wrapped `value`:

```ts
function renderParams(model: QueryBuilderOperation, def: QueryBuilderOperationDef, innerExpr: string) {
  return (model.params ?? []).map((value, index) => {
    const paramDef = def.params[index];
    if (paramDef?.type === 'string') {
      return `"${value}"`;
    }

    return value;
  });
}
```

### PR DESCRIPTION
**Pull request by lasr3**
**What is this feature?**

This feature improves error handling in Grafana's Prometheus explorer, preventing crashes from invalid queries by providing feedback. It adds optional chaining to handle missing values. Integration tests are also added to prevent future issues.

**Why do we need this feature?**

We need this feature to fix the open issue where Grafana crashes when handling invalid Prometheus queries. Instead of crashing, the editor should provide feedback on query errors.

**Who is this feature for?**

This feature is for Grafana users, especially developers and engineers working with Prometheus queries in the Explore section.

**Which issue(s) does this PR fix?**:

Fixes #94219 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Review by NWRichmond — CHANGES_REQUESTED** [review]
Thanks for this contribution, @devluisrodrigues!

A few suggestions:
- Let's only use optional chaining where required
- Please ensure that any new or changed tests run & pass
- With respect to testing, the guidance in https://github.com/grafana/grafana/issues/94219#issue-2564245262 is:
> A higher-level unit or integration test could help to guard against future regressions.

What we mean by "a higher-level unit or integration test" is a test of a function that calls `renderParams` (either directly or indirectly), or possibly a broader test of the query editor. We want to test for the expected behavior (a problematic query doesn't crash the [Query Editor in Code Mode](https://grafana.com/docs/grafana/latest/datasources/prometheus/query-editor/#code-mode)), as we are trying to move away from low-level unit tests of things like `renderParams`, which we can think of as an implementation detail. Please let me know if I can help clarify this further :). If a higher-level test is too cumbersome, I think we'd prefer to accept this PR without new tests than with a unit test of `renderParams`.

File: packages/grafana-prometheus/src/querybuilder/operationUtils.test.ts
[the remark this answers is not in the document]
Comment (devluisrodrigues) [reply packages/grafana-prometheus/src/querybuilder/operationUtils.test.ts:1]: Perfect, I misunderstood the testing part and got a little lost. I'll remove the unit test for renderParams. Thank you for your time! :)

[the remark this answers is not in the document]
Comment (devluisrodrigues) [reply packages/grafana-prometheus/src/querybuilder/operationUtils.test.ts:1]: Sorry for the missing imports, I got a little confused with my local repository. Initially, I was hesitant to add unnecessary code, but I made a mistake and forgot to add the imports at the beginning. Sorry for that. Anyway, as you mentioned preferring the PR without the unit tests, I’ve just committed a change to remove them.

**Comment by devluisrodrigues** [comment]
Hey @NWRichmond, thank you very much for taking the time to review my contribution and provide feedback.

I’ve just committed the changes you suggested, and I believe all the necessary adjustments have been made in this PR.

I’ll now do my best to simulate a high-level test.

File: packages/grafana-prometheus/src/querybuilder/operationUtils.ts [hunk 1]
```diff
@@ -95,9 +95,9 @@ export function rangeRendererLeftWithParams(
 
 function renderParams(model: QueryBuilderOperation, def: QueryBuilderOperationDef, innerExpr: string) {
   return (model.params ?? []).map((value, index) => {
-    const paramDef = def.params[index];
-    if (paramDef.type === 'string') {
-      return '"' + value + '"';
+    const paramDef = def?.params?.[index];
```
[the remark this answers is not in the document]
Comment (devluisrodrigues) [reply packages/grafana-prometheus/src/querybuilder/operationUtils.ts]: Okay, thank you for the explanation. I'll remove the unnecessary chaining operations.

**Comment by devluisrodrigues** [comment]
Hey @NWRichmond, I just built the application and tested the error:
Screenshot from 2024-10-18 07-58-08

It's working as expected!
If you need anything else, feel free to ask! :)

**Comment by NWRichmond** [comment]
Thank you for the updates! I'll review them this week :)

**Review by NWRichmond — APPROVED** [review]
Thanks for the updates, @devluisrodrigues, and thanks for contributing to Grafana!

Approving now, but will wait to merge this until next week.

**Comment by NWRichmond** [comment]
Hi @devluisrodrigues! If you could please sign the Contributor License Agreement, the rest of the CI/CD checks should work: https://cla-assistant.io/grafana/grafana?pullRequest=94421

**Comment by devluisrodrigues** [comment]
> Hi @devluisrodrigues! If you could please sign the [Contributor License Agreement](https://github.com/grafana/grafana/pull/94421#issuecomment-2400117648), the rest of the CI/CD checks should work: https://cla-assistant.io/grafana/grafana?pullRequest=94421

oh sorry, about that. I just signed the License!

**Comment by devluisrodrigues** [comment]
Sorry @NWRichmond, I pressed by mistake the request review button.

**Pull request merged**